contracts/strategies/BenqiStrategyV1.sol

```diff
@@ -0,0 +1,357 @@
+            "Swap pair 1 does not match rewardToken1"
+        );
+
+        swapPairToken0 = IPair(_swapPairToken0);
+        swapPairToken1 = IPair(_swapPairToken1);
+    }
+
+    function setAllowances() public override onlyOwner {
+        depositToken.approve(address(tokenDelegator), type(uint256).max);
+        tokenDelegator.approve(address(tokenDelegator), type(uint256).max);
+    }
+
+    function deposit(uint amount) external override {
+        _deposit(msg.sender, amount);
+    }
+
+    function depositWithPermit(uint amount, uint deadline, uint8 v, bytes32 r, bytes32 s) external override {
+        depositToken.permit(msg.sender, address(this), amount, deadline, v, r, s);
+        _deposit(msg.sender, amount);
+    }
+
+    function depositFor(address account, uint amount) external override {
+        _deposit(account, amount);
+    }
+
+    function _deposit(address account, uint amount) private onlyAllowedDeposits {
+        require(DEPOSITS_ENABLED == true, "BenqiStrategyV1::_deposit");
+        if (MAX_TOKENS_TO_DEPOSIT_WITHOUT_REINVEST > 0) {
+            uint unclaimedRewards = checkReward();
+            if (unclaimedRewards > MAX_TOKENS_TO_DEPOSIT_WITHOUT_REINVEST) {
+                _reinvest(unclaimedRewards);
+            }
+        }
+        require(depositToken.transferFrom(msg.sender, address(this), amount), "BenqiStrategyV1::transfer failed");
+        _stakeDepositTokens(amount);
+        _mint(account, getSharesForDepositTokens(amount));
+        totalDeposits = totalDeposits.add(amount);
+        emit Deposit(account, amount);
+    }
+
+    function withdraw(uint amount) external override {
+        uint depositTokenAmount = getDepositTokensForShares(amount);
+        if (depositTokenAmount > 0) {
+            uint toWithdraw = _withdrawDepositTokens(depositTokenAmount);
+            _safeTransfer(address(depositToken), msg.sender, toWithdraw);
+            _burn(msg.sender, amount);
+            totalDeposits = totalDeposits.sub(depositTokenAmount);
+            emit Withdraw(msg.sender, depositTokenAmount);
+        }
+    }
+
+    function _withdrawDepositTokens(uint amount) private returns (uint) {
+        require(amount > 0, "BenqiStrategyV1::_withdrawDepositTokens");
+        _unrollDebt();
+        uint balance = tokenDelegator.balanceOfUnderlying(address(this));
+        uint toWithdraw = balance.mul(amount).div(totalSupply);
+        uint newBalance = balance.sub(toWithdraw);
+        require(tokenDelegator.redeemUnderlying(toWithdraw) == 0, "BenqiStrategyV1::redeem failed");
+        if (newBalance > 0) {
+            _rollupDebt(newBalance, 0);
+        }
+        
+        return toWithdraw;
+    }
+
+    function reinvest() external override onlyEOA {
+        uint unclaimedRewards = checkReward();
+        require(unclaimedRewards >= MIN_TOKENS_TO_REINVEST, "BenqiStrategyV1::reinvest");
+        _reinvest(unclaimedRewards);
```

Comment:
we effectively check balances multiple times: in `checkReward` and `_reinvest`. Could be worth creating a second view for the rewards, to return `(uint avaxReward, uint qiReward, uint estWavax)` or similar and pass the reward amounts to `_reinvest` and use the estimated conversion for the check